### ISSUE TITLE
Dropdown: don't ignore keydown from input or textarea: there's a more natural solution.

### ISSUE THREAD

**Issue by lddubeau** [twbs/bootstrap#15757]
### Versions

Browser: Chrome 39
OS: Debian testing, as of 2015-02-04.
Bootstrap: 3.3.2

### What steps will reproduce the problem?

Open this bin that uses Bootstrap 3.1.1 http://jsbin.com/cilope and do:

1. Ignore "Custom Key Events" for now. Click on "Stock Key Events".

2. Click in the text field.

3. Type the down arrow on the keyboard.

The first item in the menu will be highlighted.

Now open this bin that uses 3.3.2 http://jsbin.com/horino and perform the same steps as above. The first item in the menu won't be highlighted.

### What is expected to happen?

In both cases the first item in the menu should be highlighted.

### What happens instead?

In the 2nd case (Bootstrap 3.3.2), it is not highlighted.

### Observations

The problem is with the test at the start of the `keydown` method. Specifically the 2nd part of the test that checks the `tagName` of the target of the event:

    Dropdown.prototype.keydown = function (e) {
      if (!/(38|40|27|32)/.test(e.which) || /input|textarea/i.test(e.target.tagName)) return

It seems to be an attempt at preventing keyboard events that originated in an `input` or `textarea` that is a child of the dropdown from being processed by the dropdown. But what if the developer *does* want the dropdown to process the event? In my case, I have a dropdown menu with a filter field at the start that allows narrowing the set of choices. I *do* want the user to be able to hit the down arrow key while in the field to move to menu items.

Ok, what about those people who *do not* want the event to be processed by the dropdown? They can use the time-tested and customary method of calling `stopPropagation` on the event, either explicitly or by returning `false` from their event handler if they also want to prevent default handling. There are two examples of this in the bins above. First, there's the `click` event which has a handler on the input field preventing it from propagating to the dropdown. Then there's the "Custom Key Events" dropdown which works both in 3.1.1 and 3.3.2. It captures the down arrow key, clears the field and prevents it from propagating to the dropdown. 

There's no need for `/input|textarea/i.test(e.target.tagName).`. Not only is it not needed but it complicates the lives of the developers who use Bootstrap.

**Comment by twbs-lmvtfy** [comment]
Hi @lddubeau!

You appear to have posted a live example (http://jsbin.com/suxica/edit), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 19, column 11: Element ```` input ```` not allowed as child of element ```` ul ```` in this context. (Suppressing further errors from this subtree.)
* line 19, column 19: Stray end tag ```` input ````.
* line 27, column 127: Duplicate ID ```` dropdownMenu1 ````.
* line 32, column 11: Element ```` input ```` not allowed as child of element ```` ul ```` in this context. (Suppressing further errors from this subtree.)
* line 32, column 19: Stray end tag ```` input ````.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by twbs-lmvtfy** [comment]
Hi @lddubeau!

You appear to have posted a live example (http://jsbin.com/bevonu/edit), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 19, column 11: Element ```` input ```` not allowed as child of element ```` ul ```` in this context. (Suppressing further errors from this subtree.)
* line 19, column 19: Stray end tag ```` input ````.
* line 27, column 127: Duplicate ID ```` dropdownMenu1 ````.
* line 32, column 11: Element ```` input ```` not allowed as child of element ```` ul ```` in this context. (Suppressing further errors from this subtree.)
* line 32, column 19: Stray end tag ```` input ````.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by lddubeau** [comment]
The fixed examples have been edited into the original post but I guess that does not satisfy the bot, so:

Bootstrap 3.1.1: http://jsbin.com/cilope

Bootstrap 3.3.2: http://jsbin.com/horino

**Comment by twbs-lmvtfy** [comment]
Hi @lddubeau!

You appear to have posted a live example (http://jsbin.com/horino/edit), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* [W002](https://github.com/twbs/bootlint/wiki/W002): `<head>` is missing X-UA-Compatible `<meta>` tag that disables old IE compatibility modes
* [W003](https://github.com/twbs/bootlint/wiki/W003): `<head>` is missing viewport `<meta>` tag that enables responsiveness

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by twbs-lmvtfy** [comment]
Hi @lddubeau!

You appear to have posted a live example (http://jsbin.com/cilope/edit), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* [W002](https://github.com/twbs/bootlint/wiki/W002): `<head>` is missing X-UA-Compatible `<meta>` tag that disables old IE compatibility modes
* [W003](https://github.com/twbs/bootlint/wiki/W003): `<head>` is missing viewport `<meta>` tag that enables responsiveness
* line 5, column 1: [W013](https://github.com/twbs/bootlint/wiki/W013): Bootstrap version might be outdated. Latest version is at least 3.3.2 ; saw what appears to be usage of Bootstrap 3.1.1
* line 6, column 1: [W013](https://github.com/twbs/bootlint/wiki/W013): Bootstrap version might be outdated. Latest version is at least 3.3.2 ; saw what appears to be usage of Bootstrap 3.1.1

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by lddubeau** [comment]
My response to these Bootlint issues:

1. Not using the latest version in one of the examples: well, I **am** showing the difference in handling between 3.1.1 and 3.3.2. 

2. Re X-UA-Compatible: IE compatibility modes is not what is at stake here.

3. Re responsiveness: Responsiveness is not what is at stake here.

**Comment by cvrebert** [comment]
I would argue that having Dropdowns pay attention to such events by default violates the Principle of Least Surprise and hurts composability by making the behavior context-specific. I could no longer just toss an `<input>` within a dropdown and expect the `<input>` to work the same way it does anywhere else on the page.

Having to define an event handler would be slightly annoying, and handling this somewhat obscure situation will often be overlooked by authors. Plus, we try to make the widgets mostly usable exclusively through `data-*` attributes, without the author writing any JS.

Adding some config option or magic "pay attention to this `<input>`/`<textarea>`" attribute sounds reasonable though.

**Comment by lddubeau** [comment]
On Mon, 2015-02-09 at 11:13 -0800, Chris Rebert wrote:
> I would argue that having Dropdowns pay attention to such events by
> default violates the Principle of Least Surprise and hurts
> composability by making the behavior context-specific. I could no
> longer just toss an <input> within a dropdown and expect the <input>
> to work the same way it does anywhere else on the page.

It is generally the case that one cannot just toss something that has a
behavior of its own into a dropdown and find that it will work. For
instance, I've got `<button>` elements in some of my menus. They change
the contents of the menu but do not select a menu item. Of course, I
have to prevent `click` or the user hitting the Enter key on these
buttons from reaching the dropdown, otherwise the dropdown closes.
Similarly, if someone would want to add a collapsible section in a menu,
for instance to contain options that people do not normally use (e.g.
advanced search options, or advanced login options), they'd have to do
the same kind of capture to prevent the menu from closing when the user
tries to use the mouse or the keyboard to open or close the collapsible.
Or a dropdown with tabs in it? Someone tried it: 

http://stackoverflow.com/questions/21525440/twitter-bootstrap-dropdown-with-tabs-inside

Indeed they found that the events that manipulate the tabs and would
otherwise close the dropdown have to be handled with custom event
handlers.

That `<input>` and `<textarea>` happen to work without requiring
additional event handling for some use-cases with the current version of
Bootstrap is the oddity. 

> Having to define an event handler would be slightly annoying, and
> handling this somewhat obscure situation will often be overlooked by
> authors. 

I find that the default mindset I should have when I compose Bootstrap
constructs is "watch out for event propagation." Otherwise, I get
surprises. Then again, I may be doing things with Bootstrap that most
people don't do.

> Plus, we try to make the widgets mostly usable exclusively through
> data-* attributes, without the author writing any JS.

I agree that trying to avoid custom JS is a worthy goal.

> Adding some config option or magic "pay attention to this
> <input>/<textarea>" attribute sounds reasonable though.

I could live with an option on the dropdown that would mean "don't
ignore events bubbling up from your descendants". I would prefer this to
be the default behavior for the reasons I've given above. However, if it
is not the default and I have to set an option this would be much better
than the event refiring I'm doing now.

**Comment by cvrebert** [comment]
CC: @fat

**Comment by pagameba** [comment]
I also just discovered this change in behaviour, previously my search input would open a dropdown showing search results and hitting the down arrow would move into the dropdown menu which was the intended behaviour.  I think the /input|textarea/ test is too liberal, up and down arrow events on inputs should not be ignored.

**Comment by cvrebert** [comment]
@XhmikosR @hnrch02 You guys have any opinions on this?

**Comment by hnrch02** [comment]
FWIW, we added this to allow spaces to be typed out in an input within a dropdown, see #15084 and http://jsbin.com/vurico/1.
I don't like the thought of an option for this, maybe it would suffice to just ignore spaces coming from inputs and ignore up/down/space from textareas?

**Comment by twbs-lmvtfy** [comment]
Hi @hnrch02!

You appear to have posted a live example (http://jsbin.com/vurico/1/edit), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* [W002](https://github.com/twbs/bootlint/wiki/W002): `<head>` is missing X-UA-Compatible `<meta>` tag that disables old IE compatibility modes
* [W003](https://github.com/twbs/bootlint/wiki/W003): `<head>` is missing viewport `<meta>` tag that enables responsiveness
* line 6, column 3: [W013](https://github.com/twbs/bootlint/wiki/W013): Bootstrap version might be outdated. Latest version is at least 3.3.4 ; saw what appears to be usage of Bootstrap 3.3.0

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by fat** [comment]
hm this is a hard one. I think there are a bunch of good proposals – I feel @lddubeau should definitely be able to accomplish what he's trying to do and I'd rather not have an option because i think that complicates the plugin a bit (and most people aren't putting inputs in their dropdowns to begin with) so shouldn't have to deal with that decision.

So i guess i lean slightly more towards @hnrch02's solution or a data-attribute driven solution like chris suggested

**Comment by nalmar** [comment]
I was working with combobox ( input + dropdown ) on a screen-reader / keyboard only accessibility project.

I have the same problem as @pagameba on this slightly different problem which would also be fixed with @hnrch02 solution. Just block space, not arrows and escape

http://www.bootply.com/2gHt0MWRWd

I'd definitely prefer not to have to add jQuery hack to bring back this usefull feature we had pre 3.3.1.

**Comment by twbs-lmvtfy** [comment]
Hi @nalmar!

You appear to have posted a live example (http://s.bootply.com/render/2gHt0MWRWd), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 47, column 9 thru line 47, column 34: Bad value ```` option ```` for attribute ```` role ```` on element ```` a ````.
* line 48, column 32 thru line 48, column 57: Bad value ```` option ```` for attribute ```` role ```` on element ```` a ````.
* line 49, column 9 thru line 49, column 34: Bad value ```` option ```` for attribute ```` role ```` on element ```` a ````.
* line 50, column 9 thru line 50, column 34: Bad value ```` option ```` for attribute ```` role ```` on element ```` a ````.
* line 51, column 9 thru line 51, column 34: Bad value ```` option ```` for attribute ```` role ```` on element ```` a ````.
* line 52, column 9 thru line 52, column 34: Bad value ```` option ```` for attribute ```` role ```` on element ```` a ````.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by nalmar** [comment]
I was working with combobox ( input + dropdown ) on a screen-reader / keyboard only accessibility project.

I have the same problem as @pagameba on this slightly different problem which would also be fixed with @hnrch02 solution. Just block space, not arrows and escape

http://www.bootply.com/2gHt0MWRWd ( Fixed to please the bot )

I'd definitely prefer not to have to add jQuery hack to bring back this usefull feature we had pre 3.3.1.

**Comment by twbs-lmvtfy** [comment]
Hi @nalmar!

You appear to have posted a live example (http://s.bootply.com/render/2gHt0MWRWd), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* line 41, column 1: [E012](https://github.com/twbs/bootlint/wiki/E012): `.input-group` and `.col-*-*` cannot be used directly on the same element. Instead, nest the `.input-group` within the `.col-*-*`
* line 41, column 1: [E014](https://github.com/twbs/bootlint/wiki/E014): Columns (`.col-*-*`) can only be children of `.row`s or `.form-group`s
* line 9, column 9: [W013](https://github.com/twbs/bootlint/wiki/W013): Bootstrap version might be outdated. Latest version is at least 3.3.4 ; saw what appears to be usage of Bootstrap 3.3.0
* line 61, column 9: [W013](https://github.com/twbs/bootlint/wiki/W013): Bootstrap version might be outdated. Latest version is at least 3.3.4 ; saw what appears to be usage of Bootstrap 3.3.0

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by mdo** [comment]
[**Bootstrap 3 is no longer being officially developed or supported.**](https://github.com/twbs/bootstrap/issues/20631)

All work has moved onto our next major release, v4. As such, this issue or pull request is being closed as a "won't fix." For additional help and support, we recommend [utilizing our community resources](https://github.com/twbs/bootstrap/blob/master/README.md#community). Thanks for your understanding, and see you on the other side of v4!

<3,
@mdo and team

**Comment by gpakosz** [comment]
@mdo If find it disappointing you hastily closed this issue while an interesting discussion was ongoing... Because `v4` is coming...

Except `v4` [still suffers from this exact issue](https://github.com/twbs/bootstrap/blob/3c12d73203bad27edcaa1072e263fdab2a2c14f7/js/src/dropdown.js#L214-L218).

Guess what? It's my first Bootstrap based site and first thing I tried to present my navbar's search form's results was putting an input inside a dropdown... I get that I'm a webdev amateur but that's what Bootstrap's documentation lead me to try first. 🙁
